datafusion/common/src/file_options/mod.rs

```diff
@@ -0,0 +1,269 @@
+pub enum FileTypeWriterOptions {
+    Parquet(ParquetWriterOptions),
+    CSV(CsvWriterOptions),
+    JSON(JsonWriterOptions),
+    Avro(AvroWriterOptions),
+    Arrow(ArrowWriterOptions),
+}
+
+impl FileTypeWriterOptions {
+    /// Constructs a FileTypeWriterOptions given a FileType to be written
+    /// and arbitrary String tuple options. May return an error if any
+    /// string setting is unrecognized or unsupported.
+    pub fn build(
+        file_type: &FileType,
+        config_defaults: &ConfigOptions,
+        statement_options: &StatementOptions,
+    ) -> Result<Self> {
+        let options = (config_defaults, statement_options);
+
+        let file_type_write_options = match file_type {
+            FileType::PARQUET => {
+                FileTypeWriterOptions::Parquet(ParquetWriterOptions::try_from(options)?)
+            }
+            FileType::CSV => {
+                FileTypeWriterOptions::CSV(CsvWriterOptions::try_from(options)?)
+            }
+            FileType::JSON => {
+                FileTypeWriterOptions::JSON(JsonWriterOptions::try_from(options)?)
+            }
+            FileType::AVRO => {
+                FileTypeWriterOptions::Avro(AvroWriterOptions::try_from(options)?)
+            }
+            FileType::ARROW => {
+                FileTypeWriterOptions::Arrow(ArrowWriterOptions::try_from(options)?)
+            }
+        };
+
+        Ok(file_type_write_options)
+    }
+
+    /// Constructs a FileTypeWriterOptions from session defaults only.
+    pub fn build_default(
+        file_type: &FileType,
+        config_defaults: &ConfigOptions,
+    ) -> Result<Self> {
+        let empty_statement = StatementOptions::new(vec![]);
+        let options = (config_defaults, &empty_statement);
+
+        let file_type_write_options = match file_type {
+            FileType::PARQUET => {
+                FileTypeWriterOptions::Parquet(ParquetWriterOptions::try_from(options)?)
+            }
+            FileType::CSV => {
+                FileTypeWriterOptions::CSV(CsvWriterOptions::try_from(options)?)
+            }
+            FileType::JSON => {
+                FileTypeWriterOptions::JSON(JsonWriterOptions::try_from(options)?)
+            }
+            FileType::AVRO => {
+                FileTypeWriterOptions::Avro(AvroWriterOptions::try_from(options)?)
+            }
+            FileType::ARROW => {
+                FileTypeWriterOptions::Arrow(ArrowWriterOptions::try_from(options)?)
+            }
+        };
+
+        Ok(file_type_write_options)
+    }
+
+    /// Tries to extract ParquetWriterOptions from this FileTypeWriterOptions enum.
+    /// Returns an error if a different type from parquet is set.
+    pub fn try_into_parquet(&self) -> Result<&ParquetWriterOptions> {
+        match self {
+            FileTypeWriterOptions::Parquet(opt) => Ok(opt),
+            _ => Err(DataFusionError::Internal(
+                "Expected parquet options but found options for a different FileType!"
```

Comment:
I think it would help usablity if these errors contained the name of the other type. Also should these be `DataFusionError::InvalidOption`? `DataFusionError::Internal` is supposed to signal a bug

```suggestion
                "Expected parquet options but found options for: {}", self.name()
```

The same comment applies to the other `try_into*` functions as well

Reply:
I chose internal error here because this error is only hit if somehow the calling code expects options for format A but actually options for format B were set. That should not be possible based on any input a user provides. It should only happen if we have an internal logic bug (like calling try_into_json within parquet_sink).

InvalidOption error happens if the user provides invalid options, but invalid options alone cannot lead to this particular error (at least I can't think of a way without some internal bug).